src/authentication/MobileRouter.jsx

```diff
@@ -1,39 +1,128 @@
-import React from 'react'
+import React, { Component } from 'react'
 import { Router, withRouter } from 'react-router'
-
+import Proptypes from 'prop-types'
 import Authentication from './src/Authentication'
 import Revoked from './src/Revoked'
 import { logException } from 'drive/lib/reporter'
+import {
+  readState,
+  secretExchange,
+  getAccessToken,
+  readSecret,
+  clearSecret,
+  clearState,
+  checkExchangedInformations,
+  checkIfOnboardingLogin
+} from './src/utils/onboarding'
+
+class MobileRouter extends Component {
+  async checkState(receivedState, code, cozy_url, history) {
+    const localState = await readState()
+    const localSecret = await readSecret()
+    console.log({ localSecret })
+    console.log({ localState })
+
+    try {
+      if (localState !== receivedState) {
+        console.warn('States are not equals')
+        throw new Error('ERROR', 'States are not equals')
+      }
+
+      const clientInfo = await secretExchange(
+        localSecret,
+        cozy_url,
+        this.props.client
+      )
+
+      const { onboarding_secret, onboarding_state } = clientInfo
+      console.log({ onboarding_secret })
+      console.log({ onboarding_state })
+      if (
+        !checkExchangedInformations(
+          localSecret,
+          onboarding_secret,
+          localState,
+          onboarding_state
+        )
+      )
+        throw new Error('ERROR', 'exchanged informations are not good')
 
-const MobileRouter = ({
-  history,
-  appRoutes,
-  isAuthenticated,
-  isRevoked,
-  onAuthenticated,
-  onLogout,
-  appIcon
-}) => {
-  if (!isAuthenticated) {
-    return (
-      <Authentication
-        router={history}
-        onComplete={onAuthenticated}
-        onException={logException}
-        appIcon={appIcon}
-      />
-    )
-  } else if (isRevoked) {
-    return (
-      <Revoked
-        router={history}
-        onLogBackIn={onAuthenticated}
-        onLogout={onLogout}
-      />
-    )
-  } else {
-    return <Router history={history}>{appRoutes}</Router>
+      const getTokenRequest = await getAccessToken(
+        clientInfo,
+        cozy_url,
+        code,
+        this.props.client
+      )
+      const token = await getTokenRequest.json()
+      if (getTokenRequest.status !== 200) {
+        throw new Error('ERROR', token.error)
+      }
+
+      const afterAuth = await this.props.onAuthenticated({
+        url: `https://${cozy_url}`,
+        token,
+        clientInfo,
+        router: history
+      })
+      console.log('afterAuth', afterAuth)
+      clearState()
+      clearSecret()
+    } catch (error) {
+      console.log('error', error)
+      clearState()
+      clearSecret()
```

Comment:
I don't know if I can use `finally` here since I can dispatch `LOGOUT`. And I don't know, if in this case, the `finally` will be executed. If you are sure about it, then yes I can use the finally. 

No need to `await` since I don't need their return value or something else